src/client/executor.rs

```diff
@@ -363,36 +443,112 @@ impl Client {
+                Retryability::Read if self.inner.options.retry_reads != Some(false) => {
+                    return Ok(Retryability::Read);
+                }
+                Retryability::Write if conn.stream_description()?.supports_retryable_writes() => {
+                    // commitTransaction and abortTransaction should be retried regardless of the
+                    // value for retry_writes set on the Client
+                    if op.name() == "commitTransaction"
+                        || op.name() == "abortTransaction"
+                        || self.inner.options.retry_writes != Some(false)
+                    {
+                        return Ok(Retryability::Write);
+                    }
+                }
+                _ => {}
             }
-            _ => Ok(Retryability::None),
         }
+        Ok(Retryability::None)
     }
 }
 
-/// Returns an Error with a "RetryableWriteError" label added if necessary. On a pre-4.4
-/// connection, a label should be added to any write-retryable error. On a 4.4+ connection, a
-/// label should only be added to network errors. Regardless of server version, a label should
-/// only be added if the `retry_writes` client option is not set to `false`.
-async fn get_error_with_retryable_write_label(conn: &Connection, err: Error) -> Result<Error> {
-    if let Some(max_wire_version) = conn.stream_description()?.max_wire_version {
-        if err.should_add_retryable_write_label(max_wire_version) {
-            return Ok(err.with_label("RetryableWriteError"));
+impl Error {
+    /// Adds the necessary labels to this Error.
+    ///
+    /// A TransientTransactionError label should be added if a transaction is in progress and the
+    /// error is a network or server selection error.
+    ///
+    /// On a pre-4.4 connection, a RetryableWriteError label should be added to any write-retryable
+    /// error. On a 4.4+ connection, a label should only be added to network errors. Regardless of
+    /// server version, a label should only be added if the `retry_writes` client option is not set
+    /// to `false`, the operation during which the error occured is write-retryable, and a
+    /// TransientTransactionError label has not already been added.
+    fn add_labels(
+        &mut self,
+        conn: Option<&Connection>,
+        session: &Option<&mut ClientSession>,
+        retryability: Option<&Retryability>,
+    ) -> Result<()> {
+        let transaction_state = session.as_ref().map_or(&TransactionState::None, |session| {
+            &session.transaction.state
+        });
+        match transaction_state {
+            TransactionState::Starting | TransactionState::InProgress => {
+                if self.is_network_error() || self.is_server_selection_error() {
+                    self.add_label(TRANSIENT_TRANSACTION_ERROR);
+                }
+            }
+            TransactionState::Committed { .. } => {
+                if let Some(conn) = conn {
```

Comment:
given that this is computed in a few branches, could we bring it out to a variable that's computed outside the `match`? It's pretty cheap to compute so it shouldn't be a perf issue.